otj-server-core/src/main/java/com/opentable/server/SpringPortSelectionPostProcessor.java

```diff
@@ -0,0 +1,80 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.opentable.server;
+
+import java.util.Arrays;
+import java.util.Locale;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.env.EnvironmentPostProcessor;
+import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.core.env.PropertySource;
+import org.springframework.lang.NonNull;
+
+public class SpringPortSelectionPostProcessor implements EnvironmentPostProcessor {
+
+    private static final String PORT_SELECTOR_PROPERTY_SOURCE = "OtPortSelectorPropertySource";
+
+    static final String MANAGEMENT_SERVER_PORT = "management.server.port";
+    static final String JMX_PORT = "ot.jmx.port";
+    static final String HTTPSERVER_CONNECTOR_DEFAULT_HTTP_PORT = "ot.httpserver.connector.default-http.port";
+    static final String SERVER_PORT = "server.port";
+
+    @Override
+    public void postProcessEnvironment(final ConfigurableEnvironment environment, final SpringApplication application) {
+        Arrays.stream(environment.getActiveProfiles())
+                .filter("deployed"::equals)
+                .findFirst()
+                .map(i -> environment)
+                .ifPresent(this::injectPortSelectorPropertySource);
+    }
+
+    private void injectPortSelectorPropertySource(ConfigurableEnvironment environment) {
+        environment.getPropertySources()
+                .remove(PORT_SELECTOR_PROPERTY_SOURCE);
+        environment.getPropertySources()
+                .addLast(new PropertySource<Integer>(PORT_SELECTOR_PROPERTY_SOURCE) {
+                    @Override
+                    public Integer getProperty(@NonNull String s) {
+                        boolean isK8s = (!"IS_KUBERNETES".equalsIgnoreCase(s)) && "true".equalsIgnoreCase(environment.getProperty("IS_KUBERNETES", "false"));
+                        // Default spring boot
+                        if (SERVER_PORT.equalsIgnoreCase(s)) {
+                            return Integer.parseInt(environment.getProperty("PORT_HTTP",
+                                    environment.getProperty("PORT0", "0")));
+                        }
+                        // otj-server default connector
+                        if (HTTPSERVER_CONNECTOR_DEFAULT_HTTP_PORT.equalsIgnoreCase(s)) {
+                            return Integer.parseInt(environment.getProperty("PORT_HTTP",
+                                    environment.getProperty("PORT0", "-1")));
+                        }
+                        // otj-server named connector
+                        if (isK8s && s.matches("ot\\.httpserver\\.connector\\..*-.*\\.port")) {
+                            final String name = s.split("\\.")[3].split("-")[0].toUpperCase(Locale.US);
+                            return Integer.parseInt(environment.getProperty("PORT_" + name, "-1"));
```

Comment:
The correct thing to do is read PORT_$PROTOCOL. I recognize this makes it tricky to have more than one there, but otherwise most of the time this fails

protocol should be one of http, https, proxy+http, or proxy+https

Reply:
Why does PORT_MY work

Reply:
why do you split on "-" instead of ."